frame/contracts/src/lib.rs

```diff
@@ -1081,40 +1145,54 @@ impl<T: Trait + Send + Sync> SignedExtension for CheckBlockGasLimit<T> {
 	type Call = <T as Trait>::Call;
 	type AdditionalSigned = ();
 	type DispatchInfo = DispatchInfo;
-	type Pre = ();
+	/// Optional pre-dispatch check data.
+	///
+	/// It is present only for the contract calls that operate with gas.
+	type Pre = Option<DynamicWeightData<T::AccountId, NegativeImbalanceOf<T>>>;
+
+	fn additional_signed(&self) -> Result<(), TransactionValidityError> { Ok(()) }
 
-	fn additional_signed(&self) -> sp_std::result::Result<(), TransactionValidityError> { Ok(()) }
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		_: DispatchInfo,
+		_: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		Self::perform_pre_dispatch_checks(who, call)
+	}
 
 	fn validate(
 		&self,
-		_: &Self::AccountId,
+		who: &Self::AccountId,
 		call: &Self::Call,
 		_: Self::DispatchInfo,
 		_: usize,
 	) -> TransactionValidity {
-		let call = match call.is_sub_type() {
-			Some(call) => call,
-			None => return Ok(ValidTransaction::default()),
-		};
+		Self::perform_pre_dispatch_checks(who, call)
+			.map(|_| ValidTransaction::default())
+	}
 
-		match call {
-			Call::claim_surcharge(_, _) | Call::update_schedule(_) =>
-				Ok(ValidTransaction::default()),
-			Call::put_code(gas_limit, _)
-				| Call::call(_, _, gas_limit, _)
-				| Call::instantiate(_, gas_limit, _, _)
-			=> {
-				// Check if the specified amount of gas is available in the current block.
-				// This cannot underflow since `gas_spent` is never greater than `T::BlockGasLimit`.
-				let gas_available = T::BlockGasLimit::get() - <Module<T>>::gas_spent();
-				if *gas_limit > gas_available {
-					// gas limit reached, revert the transaction and retry again in the future
-					InvalidTransaction::ExhaustsResources.into()
-				} else {
-					Ok(ValidTransaction::default())
-				}
-			},
-			Call::__PhantomItem(_, _)  => unreachable!("Variant is never constructed"),
+	fn post_dispatch(pre: Self::Pre, _info: DispatchInfo, _len: usize) {
+		if let Some(DynamicWeightData {
+			transactor,
+			imbalance,
+		}) = pre {
+			let (gas_left, gas_spent) = GasUsageReport::take();
```

Comment:
Where is this value coming from and for what does it account for? just the contract ops? or the inner calls to runtime as well?

Reply:
This value is saved in the `execute_wasm` function. It basically the total consumed gas: contract ops, inner calls to runtime, and everything else.